text/0000-interceptors.md

```diff
@@ -0,0 +1,134 @@
+
+# Motivation
+[motivation]: #motivation
+
+Currently sources can define a list of `preprocessors`, they take a `&[u8]` and produce a `Vec<Vec<u8>>`.
+Sinks can define a list of `postprocessors`, they take a `&[u8]` and produce a `Vec<Vec<u8>>`.
+When chained for each output `Vec<u8>` the next processor in the chain is called. In the source after preprocessors are done, each `Vec<u8>` is fed to a `Codec` that creates an `Value` from it. Before the `postprocessor` chain in a sink, a `Codec` creates an `Option<Vec<u8>>` from every event `Value`.
+
+A `Codec` is stateless and bidirectional. All existing pre- and postprocessors define the same operation (e.g. split a given byte sequence into lines). Interceptors would join those dual operations into one entity.
+
+It might be interesting at some point to not only allow a simple chain of interceptors, feeding bytes through them, but allow graphs with branching and joining, decisions where to continue based on investigation of the current chunk.
+
+Also pre- and postprocessors currently are not configurable, so we have different lines processors, differing only in the delimiter. This is undesirable. Thus interceptors need to be configurable.
+
+It is currently unclear how batched events are handled in sinks. Some treat each element of a batch as a single outbound unit (e.g. kafka sink where each element becomes its own message), some group batch elements in one outbound unit (e.g. elastic sink which builds up 1 request for a batch).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Interceptors take a single byte sequence, and return zero or more byte sequences. They are stateful. So, they are able to split and join byte sequences.
+They are used to handle bytes in two different directions:
+
+* From the wire to the dispatch point, where each byte sequence is decoded and turned into an event
+* From the dispatch point (from a pipeline) to the wire, where each byte sequence is put on the wire, be it into a TCP stream or as UDP packet or as HTTP request body.
+
+Interceptors are intended for tasks like splitting an incoming byte-sequence into chunks at some delimiter, compression like gzip, lz4, snappy etc., encodings like base64, handle length-prefix framing and other framing types.
+
+They are configured on sinks and sources as interceptor chains. One chain for each direction. 
+
+The `in` chain is going from the wire, starting at the first element in the list, to the dispatch point, after the last element of the list. 
+
+The `out` chain is going from the dispatch point, receiving data from the tremor-runtime, the pipeline, starting at the first element of the list, to the wire, after the last element of the `out` list.
+
+Example configuration (with imaginmary interceptors):
+
+```yaml
+source:
+  - id: ws-in
+    type: ws
+    interceptors:
```

Comment:
If we keep in and out separated what's the difference to using pre- and postprocesssors? i.e. what is the benefit we'd get from interceptors if a user still has to configure the chains separately. It seems to just move the current system into a more nested format.

Reply:
So, i first went with only 1 chain. Why this isn't satisfying, there are two elements to this:
* There is the minor issue, that users might not always want to have the exact reversed operations. Using interceptors that are strictly bidirectional and reverse their operation from the other direction, we take that possibility from users.
* Also, there more general and important (to me) is that interceptor operations are fundamentally different for the two directions. They need to be configured differently. Putting this into the same entity, conflates and confuses
The most obvious example is the interceptor splitting input bytes by `\n`. The reverse operation on outgoing data could be appending a `\n` to each incoming byte array. But this is not doing exactly the reverse, it should also join those lines, to be the exact opposite. In our current implementation, some sinks do batch, some dont.
Adding this joining/batching to the interceptor who splits in one direction and joins in the other, would require to configure the max batch size and/or match batch wait time. To me it is cleaner and more flexible to separate it out into two chains like described in the RFC. Each interceptor (`split` and `join`) has a cleaner operation, that does different things.

### benefits compared to pre-/postprocessors

* We have a clearer api. we have one entity for which the direction of processing is irrelevant.
* Clearer purpose for each interceptor due to more generic concept/operation. `split` and `join` instead of `lines`.
* They are configurable
* Having two chains does not block people who want to do different operations on incoming and outgoing data.

If you have a good idea, to smunch both chains into one and stuff is still clear and maintains the level of flexibility this RFC exposes, I am all for it.